### ISSUE TITLE
[P1][ARU][S00-ML] Second-stage per-class cap invalidates the case-control inverse-probability weights

### ISSUE THREAD

**Issue by SzeChunYiu** [SzeChunYiu/ccb-testbeam#1112]
## Evidence state / severity

- **Severity:** P1 because the producer explicitly labels this as an implementation-consistency check, not a scientific benchmark; P0 only if any downstream/public claim interprets its weighted calibration/prevalence metrics scientifically.
- **Evidence state:** **CONFIRMED from current-main sampling algebra and code path.**
- **Audited source:** `scripts/01_build_pulse_table_from_root.py` current main.
- **Parents:** S00 deterministic selector (#1109), event/run bootstrap identity, run split.

## Atomic input/output contract

The S00 ML diagnostic uses a two-stage sampling design.

### Stage 1: case-control Bernoulli sampling

For every stave-record,

```text
p1(selected)=0.20
p1(rejected)=0.05
```

by default. Each retained row is assigned

```text
sampling_weight = 1/p1(class)
```

so selected rows receive weight 5 and rejected rows weight 20. The code/documentation states that these inverse-probability weights restore population prevalence.

### Stage 2: per-class cap

After concatenating all runs, current `scan_raw()` groups `ml_rows` by `selected`, then draws at most

```text
max_sample = max_train_per_class + max_test_per_class = 60000 + 30000 = 90000
```

rows **separately from each class** using `subset.sample(n=min(len(subset),90000))`.

The existing `sampling_weight` is not changed after this second sampling stage.

`run_ml_check()` later treats that unchanged field as the population-restoring weight for training/evaluation and reports a weighted prevalence, weighted accuracy, Brier score, ROC AUC and average precision.

## Mathematical defect

Let class `c` have population size `N_c`. Stage 1 retains each row with probability `p1_c`. Conditional on the realized Stage-1 sample containing `n1_c` rows, the cap retains a row with probability

```text
p2_c = min(1, 90000/n1_c)
```

under the uniform within-class cap.

The realized inclusion probability is therefore

```text
π_c = p1_c * p2_c
```

and a design weight that reconstructs the original population must account for **both** stages, e.g. conditionally

```text
w_c ∝ 1/(p1_c*p2_c),
```

not merely `1/p1_c`.

The second-stage cap is class-specific, so omitting `p2_c` is not a harmless common normalization.

The canonical selected count alone is 640,737. With a 20% Stage-1 keep rate the expected selected ML sample is about 128,147 rows, already above the 90,000 cap. Thus the selected class is expected to undergo the unrecorded second-stage thinning in the canonical run. The same must be measured exactly for the rejected class and by train/test run stratum.

If both classes hit the same 90k cap, the final combined capped sample contains equal class counts while retaining weights 5 vs 20; its design-weighted prevalence before run splitting is mechanically driven toward

```text
(90000*5)/(90000*5 + 90000*20) = 0.20,
```

which is a property of the sampling/cap design, not necessarily the true four-stave-record prevalence.

## Competing valid designs

H1. **No second-stage cap:** retain Stage-1 case-control rows and use `1/p1_c` weights.

H2. **Two-stage design-weighted cap:** retain the cap but record exact second-stage inclusion probabilities and use total inclusion weights.

H3. **Fixed stratified sample for algorithmic diagnostics:** intentionally choose fixed class/run counts and stop claiming population-restored prevalence/calibration; report only conditional diagnostic metrics.

H4. **Direct population evaluation:** train on a bounded sample but evaluate on the full held-out run population, eliminating the need for case-control corrections in final diagnostic metrics.

H1–H4 answer different questions and should have distinct output labels.

### 1. Sampling/design-statistics lead
Specify the estimand for every reported metric: population prevalence, class-conditional discrimination, or implementation-consistency on an engineered case-control sample. Derive the inclusion probability for the chosen design.

### 2. Adversarial reviewer
Vary Stage-1 keep probabilities and per-class cap while holding the underlying synthetic population/model fixed. Any purported population metric should remain invariant within Monte Carlo error after correct weighting. The current implementation is expected to fail when caps bind asymmetrically.

### 3. Independent validation reviewer
Build exact finite-population fixtures where every row and inclusion probability is known. Compare full-population truth with Stage-1-only, current two-stage, corrected two-stage, and full-heldout evaluation. Cluster bootstrap should resample the correct target population/weighted estimator, not silently bootstrap the capped convenience sample as though it were the population.

### 4. Claims/provenance reviewer
Keep this branch labelled `implementation consistency`, as the code already intends. Do not use weighted reliability/prevalence plots as evidence of selector physics or detector performance unless sampling closure is demonstrated. Persist both sampling stages and realized counts in the manifest.

## Discriminating experiments / known-answer tests

1. Finite synthetic population with known class prevalence and deterministic features.
2. Choose Stage-1 rates 0.20/0.05 and caps low enough to bind both classes; verify corrected weighted prevalence equals truth.
3. Make the cap bind only one class; this is the sharpest test that missing `p2` cannot be absorbed into a common weight factor.
4. Repeat with different caps and seeds; population metrics must be stable.
5. Compare train-bounded / **full-heldout evaluation** against weighted capped-heldout evaluation.
6. Publish realized counts by `(run, selected)` before Stage 1, after Stage 1 and after cap.
7. Negative control: when no cap binds, corrected design must reduce exactly to current Stage-1 weights.

## Implementation plan

Preferred scientific design: bounded training sample + full held-out evaluation if memory/runtime permits.

Otherwise:

1. Attach explicit sampling-stage columns/provenance: `p_case_control`, `p_cap_conditional`, `pi_total`, `design_weight`.
2. Compute and record realized Stage-1 counts by class (and by run if capping is stratified).
3. Apply caps using an explicitly declared stratum; do not let global class capping create undocumented run-composition effects.
4. Recompute evaluation weights from the complete inclusion design.
5. Separate training weights from evaluation/design weights; calibration-model fitting may require its own strategy.
6. Store sampling design, rates, caps, realized counts and seed in the manifest.
7. Add a test proving population-prevalence recovery under cap binding.

## Acceptance criteria

- [ ] Every retained evaluation row has a documented total inclusion probability for all random subsampling stages, or evaluation occurs on the full held-out population.
- [ ] Known-answer finite-population tests recover prevalence and weighted metrics when caps bind one or both classes.
- [ ] Realized pre/post sampling counts are published by class and run.
- [ ] Changing convenience cap does not materially change population estimands after correct design adjustment.
- [ ] Output labels distinguish population metrics from engineered case-control diagnostics.
- [ ] Cluster/bootstrap uncertainty uses the declared estimator and sampling design.

## Rejection criteria

Do not close by noting that the cap is uniform *within* each class. The omitted factor differs **between classes** whenever their realized cap fractions differ, which is exactly what matters for prevalence/calibration metrics.

## Cross-scale compatibility

`deterministic selector → finite record population → case-control inclusion → class cap → run split → model fit → weighted held-out metrics → implementation-consistency report`.

This issue does not question the exact 640,737 deterministic selector count itself; it owns only the secondary ML diagnostic sampling measure.

**Comment by SzeChunYiu** [comment]
Additional current-main environment-dependent branch to include in this sampling atom: `run_ml_check()` calls `CalibratedClassifierCV.fit(..., sample_weight=w_train)` but catches `TypeError` and silently refits **without weights** on older scikit-learn. The reliability plot similarly tries `calibration_curve(..., sample_weight=sw)` and on `TypeError` silently recomputes it unweighted. Calibration/probability estimands are prevalence-sensitive, so dropping design weights is not equivalent to a harmless compatibility fallback. Acceptance should therefore require: (a) record sklearn version and whether weighted fit/calibration was actually supported/applied; (b) fail closed or explicitly label a different estimand instead of silently dropping weights; (c) known-answer test where case-control prevalence differs strongly from population prevalence, proving the weighted and unweighted calibration branches are distinguishable; and (d) one canonical implementation path for authorising artifacts.

**Comment by SzeChunYiu** [comment]
Closing under campaign ledger evidence (#1218 compliant manual close).

- Status: FIXED (fail-closed contract / ADR landed)
- Merged PR(s): #1242
- Ledger: `reports/mc_validation/execution/ISSUE_CAMPAIGN_LEDGER.md`

If this was intended as physics-BLOCKED rather than code-fixed, reopen and mark BLOCKED with ADR.
